src/credentials/GroupDataProviderImpl.h

```diff
@@ -143,9 +143,10 @@ class GroupDataProviderImpl : public GroupDataProvider
     class GroupKeyContext : public Crypto::SymmetricKeyContext
     {
     public:
-        GroupKeyContext(GroupDataProviderImpl & provider) : mProvider(provider) {}
+        GroupKeyContext(GroupDataProviderImpl & provider, bool allocated = false) : mProvider(provider) {}
```

Comment:
This needs to set `mAllocated`, no?

Reply:
Not really, but still somewhat wrong. Since this class is protected, it is always instantiated within GroupDataProviderImpl, but in two different contexts: This constructor is used by the Iterator, which holds a single instance of GroupKeyContext, allocated in the stack, so `allocated` must be set to false. The second constructor is used to to return allocated instances of GroupKeyContext to an external caller. In this case, `allocated` should always be true, which makes the second constructor wrong.
We could always allocate all instances of GroupKeyContext, avoiding any potential confusion, but since the iterator itself is allocated, I chose not to.
All thinks considered, I have removed the mAllocated flag altogether. Any external caller will receive an allocated instance of GroupKeyContext, and must call the `Release` method to release it. The only not allocated instance is private to the iterator, which simply do not call `Release`.